### PR TITLE
`multi_margin_loss`: check `weight` shape, make contiguous on CPU, add tests

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #104578
* __->__ #104852
* #104851
* #104850

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Summary:
- There was no shape check for `weight`, so this could access arbitrary memory.
- On CPU, `weight` wasn't `contiguous` unlike on CUDA. No one noticed because there were no tests for `weight`.
- Added tests for `weight`.

File: aten/src/ATen/native/LossMultiMargin.cpp [hunk 1]
```diff
@@ -125,13 +125,17 @@ void multi_margin_loss_out_cpu_template(
 
   auto input_contiguous = input.contiguous();
   auto target_contiguous = target.contiguous();
+  Tensor weight_contiguous;
+  if (weight && weight->defined()) {
+    weight_contiguous = weight->contiguous();
+  }
```
Comment (nkaretnikov) [inline aten/src/ATen/native/LossMultiMargin.cpp:131]: For compat with `multi_margin_loss_cuda_out`:

```cpp
  Tensor weights;
  if (weights_ && weights_->defined()) {
    weights = weights_->contiguous();
  }
```

File: aten/src/ATen/native/LossMultiMargin.cpp [hunk 2]
```diff
@@ -103,15 +103,15 @@ void multi_margin_loss_out_cpu_template(
     const Tensor& target,
     int p,
     const Scalar& margin,
-    const Tensor& weight,
+    const c10::optional<Tensor>& weight,
```
Comment (nkaretnikov) [inline aten/src/ATen/native/LossMultiMargin.cpp:106]: `multi_margin_loss_shape_check` accepts an optional `Tensor` because `weight` is an optional on CUDA. So I change this (and elsewhere) to be consistent.

File: aten/src/ATen/native/LossMultiMargin.cpp [hunk 3]
```diff
@@ -262,12 +266,9 @@ Tensor multi_margin_loss_cpu(
...
     const Tensor& target,
     const Scalar& p,
-    const Scalar& margin, const c10::optional<Tensor>& weight_opt,
+    const Scalar& margin,
+    const c10::optional<Tensor>& weight,
     int64_t reduction) {
-  // See [Note: hacky wrapper removal for optional tensor]
-  c10::MaybeOwned<Tensor> weight_maybe_owned = at::borrow_from_optional_tensor(weight_opt);
-  const Tensor& weight = *weight_maybe_owned;
-
```
Comment (nkaretnikov) [inline aten/src/ATen/native/LossMultiMargin.cpp:270]: Here and elsewhere: to have a consistent API for `multi_margin_loss_shape_check` (it accepts optional `weight` on CPU and CUDA).

File: aten/src/ATen/native/cuda/MultiMarginLoss.cu [hunk 4]
```diff
@@ -150,6 +151,12 @@ void multi_margin_loss_shape_check(
         target.dim() <= 1 && target.numel() == nframe,
         "inconsistent target size, expected ", nframe, " but got ",
         target.sizes());
+    if (weight && weight->defined()) {
+      TORCH_CHECK(
+          weight->dim() <= 1 && weight->numel() == dim,
+          "inconsistent weight size, expected ", dim, " but got ",
+          weight->sizes());
+    }
```
Comment (nkaretnikov) [inline aten/src/ATen/native/cuda/MultiMarginLoss.cu:159]: From the docs: "a manual rescaling weight given to each class. If given, it has to be a Tensor of size C. Otherwise, it is treated as if having all ones." But `target` here can be a scalar, so could it be useful to allow scalars for `weight`?

https://pytorch.org/docs/stable/generated/torch.nn.MultiMarginLoss.html

File: torch/_meta_registrations.py [hunk 5]
```diff
@@ -337,6 +337,11 @@ def _multi_margin_loss_shape_check(ndims, input, target):
         target.dim() <= 1 and target.numel() == nframe,
         lambda: f"inconsistent target size, expected {nframe} but got {target.shape}",
     )
+    if weight is not None:
+        torch._check(
+            weight.ndim <= 1 and weight.numel() == dim,
```
Comment (nkaretnikov) [inline torch/_meta_registrations.py:342]: Nit: use `ndim`.

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1405,6 +1406,7 @@ def sample_inputs_multi_margin_loss(op_info, device, dtype, requires_grad, **kwa
         ((S, M), make_target([S], low=0, high=M), {"margin": 1.0}),
         ((S, M), make_target([S], low=0, high=M), {"margin": -3.14}),
         ((M, S), make_target([M], low=0, high=S), {"weight": None}),
+        ((M, S), make_target([M], low=0, high=S), {"weight": make_weight([S], low=-10., high=10.)}),
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:1409]: `weight` wasn't tested before at all.

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -1427,10 +1430,11 @@ def reference_inputs_multi_margin_loss(op_info, device, dtype, requires_grad, **
     )
     ps = (1, 2)
     margins = (0, 7, -3.14)
+    weights = (None, make_weight([S], low=-10., high=10.))
     reductions = (None, "none", "mean", "sum")
 
-    for (input_shape, target), p, margin, reduction in product(inputs, ps, margins, reductions):
-        kwargs = {"p": p, "margin": margin}
+    for (input_shape, target), p, margin, weight, reduction in product(inputs, ps, margins, weights, reductions):
+        kwargs = {"p": p, "margin": margin, "weight": weight}
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:1437]: More tests with different combinations of arguments.

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -1454,6 +1458,13 @@ def error_inputs_multi_margin_loss(op, device, **kwargs):
     # invalid target dtype
     yield ErrorInput(SampleInput(make_input(5, 4), args=(make_input(5,),), kwargs={}),
                      error_type=RuntimeError, error_regex='expected scalar type Long but found Float')
+    # invalid weight
+    yield ErrorInput(SampleInput(make_input(5, 4), args=(make_input(5,),), kwargs={'weight': make_input(())}),
+                     error_type=ValueError, error_regex='weight must be one-dimensional')
+    yield ErrorInput(SampleInput(make_input(5, 4), args=(make_input(5,),), kwargs={'weight': make_input(5, 4)}),
+                     error_type=ValueError, error_regex='weight must be one-dimensional')
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:1465]: This "one-dimensional" error is only on the Python side. See [my comment above](https://github.com/pytorch/pytorch/pull/104852/files#r1257572905), maybe worth removing this restriction from the Python side and allow scalars?

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -1454,6 +1458,13 @@ def error_inputs_multi_margin_loss(op, device, **kwargs):
...
+    yield ErrorInput(SampleInput(make_input(5, 4), args=(make_input(5,),), kwargs={'weight': make_input(5,)}),
+                     error_type=RuntimeError, error_regex=r'inconsistent weight size, expected 4 but got \[5\]')
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:1467]: Raises an error now.

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -12806,6 +12809,13 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
         sample_inputs_func=sample_inputs_multi_margin_loss,
         reference_inputs_func=reference_inputs_multi_margin_loss,
         error_inputs_func=error_inputs_multi_margin_loss,
+        decorators=(
+            DecorateInfo(
+                toleranceOverride({torch.float32: tol(atol=1e-4, rtol=1e-4)}),
+                "TestJit",
+                "test_variant_consistency_jit",
+            ),
+        ),
```
Comment (nkaretnikov) [inline torch/testing/_internal/common_methods_invocations.py:12818]: Tests now multiply by weight, which affects accuracy here.

**Review by ezyang — APPROVED** [review]
